### PR TITLE
Test utils

### PR DESCRIPTION
**Pull request by anoukb27**
Added tests for generate_grid_in_polygon & group_population

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -54,4 +54,3 @@ build-backend = "poetry.core.masonry.api"
 
 [tool.pytest.ini_options]
 pythonpath = "gpbp"
-addopts = "--cov=gpbp"
```
Comment (operte) [inline pyproject.toml:57]: I also had to remove this line in order to run the debugger, otherwise it wouldn't work. But I decided not to commit this change because I wasn't sure how this affects the code coverage reports.

File: gpbp/utils.py [hunk 2]
```diff
@@ -10,6 +10,12 @@ def generate_grid_in_polygon(
     """
     This Function generates evenly spaced points within the given GeoDataFrame.
     The parameter 'spacing' defines the distance between the points in coordinate units.
+
+    AnoukB dec '24:
```
Comment (operte) [inline gpbp/utils.py:14]: I don't think it's necessary to add your name and date. One can always check who added this content in the git history.

File: gpbp/utils.py [hunk 3]
```diff
@@ -10,6 +10,12 @@ def generate_grid_in_polygon(
...
+    Function takes outer boundaries of MultiPolygon, creates a grid of points with spacing between them that
+    includes the lower boundary and excludes the upper boundary.
+    The generated grid is then clipped together with the original Multipolygon and only those points that are on the
+    grid and within the MultiPolygon are kept and returned.
```
Comment (operte) [inline gpbp/utils.py:18]: This extra explanation in the doctring is quite nice, thanks :)

File: tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+polygon1 = Polygon([(0, 0), (1, 0), (1, 1), (0, 1), (0, 0)])
+polygon2 = Polygon([(2, 2), (3, 2), (3, 3), (2, 3), (2, 2)])
+multi_polygon = MultiPolygon([polygon1, polygon2])
+
+
+class TestGenerateGridInPolygon:
+    def test_bounds_in_polygon(self):
+        grid = generate_grid_in_polygon(1, multi_polygon)
+        assert multi_polygon.bounds == (0.0, 0.0, 3.0, 3.0)
```
Comment (operte) [inline tests/test_utils.py:17]: Why did you add this assert? It doesn't seem to be related to the function you are testing. Does this serve as information for the person reading the tests?

File: tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import pandas as pd
+from more_itertools.recipes import grouper
+
+from gpbp.utils import generate_grid_in_polygon, group_population
+from shapely.geometry import Polygon, MultiPolygon
+
+
+polygon1 = Polygon([(0, 0), (1, 0), (1, 1), (0, 1), (0, 0)])
+polygon2 = Polygon([(2, 2), (3, 2), (3, 3), (2, 3), (2, 2)])
+multi_polygon = MultiPolygon([polygon1, polygon2])
```
Comment (operte) [inline tests/test_utils.py:11]: I think it would be best to transfor these, as well as `df` into fixtures. The reason being that if by any chance the functions we're testing try to change these variables, then they would affect the behaviour of other unit tests. It doesn't seem to be happening in this case, but I think it is good practice.

File: tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        grouped_pop_3 = group_population(df, 3)
+        grouped_pop_4 = group_population(df, 4)
+        assert grouped_pop_1.shape[0] == 1
+        assert grouped_pop_2.shape[0] == 2
+        assert grouped_pop_3.shape[0] == 3
+        assert grouped_pop_4.shape[0] == 4
+        assert grouped_pop_1.loc[(grouped_pop_1['longitude'] == 6.9) & (grouped_pop_1['latitude'] == 53.1)]['population'].values[0] == 14
+        assert grouped_pop_2.loc[(grouped_pop_2['longitude'] == 6.88) & (grouped_pop_2['latitude'] == 53.06)][
+                   'population'].values[0] == 12
+        assert grouped_pop_3.loc[(grouped_pop_3['longitude'] == 6.876) & (grouped_pop_3['latitude'] == 53.062)]['population'].values[0] == 9
```
Comment (operte) [inline tests/test_utils.py:69]: Is it relevant to also have a test for `grouped_pop_4`?

File: tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+df = pd.DataFrame({"longitude": [6.87641, 6.87644, 6.87964, 6.88710], "latitude": [53.06167, 53.06180, 53.06000, 53.08787],
+                   "population": [5, 4, 3, 2]})
+
+class TestGroupPopulation:
+    def test_group_pop(self):
+        grouped_pop_1 = group_population(df, 1)
+        grouped_pop_2 = group_population(df, 2)
+        grouped_pop_3 = group_population(df, 3)
+        grouped_pop_4 = group_population(df, 4)
```
Comment (operte) [inline tests/test_utils.py:61]: In these cases it's often more common to use `@pytest.mark.parametrize` to simplify the function. Have a look at Philipp's PR, I think he used it there. You could use it here as well if you would like to try it out.

File: gpbp/utils.py (hunk 2 above)
Comment (dcossao) [reply gpbp/utils.py:14]: > I don't think it's necessary to add your name and date. One can always check who added this content in the git history.

We agreed to do this during standup so we can easily differentiate our comments from ABW’s. The idea is that the comments will help future selves when refactoring, after which they will be removed.

File: pyproject.toml (hunk 1 above)
Comment (dcossao) [reply pyproject.toml:57]: If we remove this line, then the coverage report doesn’t get automatically generated when you run tests. 

Since it’s hindering development, let’s just remove it. We can run the report in other ways and/or add this line in the future.

File: gpbp/utils.py (hunk 2 above)
Comment (operte) [reply gpbp/utils.py:14]: Ah apologies, I didn't know about this. Thanks :)

File: tests/test_utils.py (hunk 4 above)
Comment (anoukb27) [reply tests/test_utils.py:17]: You are right. I was confused at first why the minimum was included and the maximum was excluded, so was asserting both the bounds as well as the maximum. But then I figured this was intended behavior (I suppose) and added it in the description of the function. No need indeed to test the bounds as they have nothing to do with the function.

File: pyproject.toml (hunk 1 above)
Comment (dcossao) [reply pyproject.toml:57]: @pkollenda: FYI

**Comment by operte** [comment]
This branch was recreated on the ABW repo [here](https://github.com/Analytics-for-a-Better-World/Public-Infrastructure-Service-Access/tree/test_utils). A new PR will be created there.
